wireapp/coturn
# WPB-10505 Fix: Rate limit by IP address to mitigate reflection attacks

----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

Coturn is vulnerable to being part of reflection/amplification based DDoS attacks by abusing the 401 unauthorized response. An attacker can spoof their source causing Coturn to reply to a target server. Using multiple instances of Coturn an attacker can create an enough traffic to take down their target. This is similar to the way NTPd has been used in DDoS attacks.

### Causes (Optional)

There are no mechanisms in place to limit the amount of 401 unauthorized responses to a specific source.

### Solutions

Implemented thread safe rate limiting for 401 Unauthorized responses:

- Rate limiting only impacts the 401 Unauthorized responses, other requests are processed normally even if an address is currently rate limited. This is intended to prevent abuse of the rate limit mechanisms
- Used native Coturn methods for storing the rate limit map via `ur_addr_map`
- Coturn native `ur_addr_map` functions have been extended to include `no_port` methods. This is because outbound port numbers are randomized, which prevents matching IPv4/IPv6 addresses between sessions.
- When the rate limit thresholds have been hit Coturn is instructed to not send a 401 Unauthorized response, but Coturn will still process the request as normal
- When `ADDR_MAP_SIZE` is hit the rate limit map is pruned for entries that have expired

### Dependencies (Optional)

N/A

Needs releases with:

- N/A

### Testing

#### Test Coverage (Optional)


#### How to Test

The tests in `examples/run_tests.sh` have been updated to include "Running rate limit by IP" and "Running NOT rated limit by IP"

Additionally one can hit the rate limit manually with the following commands:

1. Start the server: `bin/turnserver --use-auth-secret  --static-auth-secret=secret --realm=north.gov --allow-loopback-peers --no-cli --cert examples/ca/turn_server_cert.pem --pkey examples/ca/turn_server_pkey.pem`
2. Run the client four times with in a 100 second period of time: `bin/turnutils_uclient -y 127.0.0.1`

After the 4th run there should be INFO messages on the server like this example:

`16: (4117930): INFO: 401 rate limit exceeded from 127.0.0.1, response not sent`

To test if the rate limit period has timed out:

1. Kill the `turnutils_uclient`
2. Wait 100 seconds
3. Run the client: `bin/turnutils_uclient -y 127.0.0.1`

If successful the sever will create the following INFO log message:

`16: (4117930): INFO: session 009000000000000001: realm <example.com> user <>: incoming packet message processed, error 401: Unauthorized`

### Notes (Optional)

N/A

### Attachments (Optional)

N/A

----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
